lib/Sema/CSSimplify.cpp

```diff
@@ -6080,10 +6090,33 @@ ConstraintSystem::simplifyApplicableFnConstraint(
   ConstraintLocatorBuilder outerLocator =
     getConstraintLocator(anchor, parts, locator.getSummaryFlags());
 
-  // Before stripping optional types, save original type for handling
-  // @dynamicCallable applications. This supports the fringe case where
-  // `Optional` itself is extended with @dynamicCallable functionality.
-  auto origType2 = desugar2;
+  // Handle applications of types with `callAsFunction` methods.
+  // Do this before stripping optional types below, when `shouldAttemptFixes()`
+  // is true.
+  auto hasCallAsFunctionMethods =
+      desugar2->mayHaveMembers() &&
+      llvm::any_of(lookupMember(desugar2, DeclName(ctx.Id_callAsFunction)),
+                   [](LookupResultEntry entry) {
+                     return isa<FuncDecl>(entry.getValueDecl());
+                   });
+  if (hasCallAsFunctionMethods) {
+    auto memberLoc = getConstraintLocator(
```

Comment:
I'm not sure that usages of `memberLoc` or `locator` are correct/ideal here.

@xedin: could you please check these `ConstraintLocator *` usages?
You suggested [adding an "implicit call" path element](https://github.com/apple/swift/pull/24299#discussion_r298708877), but I'm not sure where to use it - please advise!

Reply:
I think for `addValueMemberConstraint` it makes more sense to use `memberLoc` rather than `locator`, and for `ApplicableFunction` indeed `locator` is best.

Regarding `implicit call` you don't have to worry about it right now, we can do it in the follow-up instead. I was thinking that it would be great to associate such new element with locator which belongs to `ApplicableFunction` something like `<call anchor> -> implicit call -> apply function` so when argument conversions are expanded by `matchCallArguments` they'd get `<call anchor> -> implicit call -> apply arg-to-param -> comparing #N to #M` locators.